py/objarray.h

```diff
@@ -38,6 +38,12 @@ typedef struct _mp_obj_array_t {
     // free is number of unused elements after len used elements
     // alloc size = len + free
     size_t free : (8 * sizeof(size_t) - 8);
+    // But for memoryview, 'free' is reused as offset (in elements) into the
+    // parent object, so provide a convinience alias for that. (Union is not
+    // used to not go into a complication od union-of-bitfields.)
+    // See comments in objarray.c for more info about memoryview
+    // representation.
+    #define memview_offset free
```

Comment:
IMO this is too hacky.  Much neater to have an accessor macro, something like `#define MP_OBJ_ARRAY_MVOFF(x) (x)->free`

Reply:
What's exactly so hacky? If it wasn't a bitfield, then just a union would have been used, with syntax like `->memview_offset`. This define achieves the same effect, except doesn't try to explore how different compilers deal with unions of bitfields (one can easily imagine that there're warts with that in different compilers).

Otherwise, compare `memview_offset` and `MVOFF`, and mind that it's used for assignment too, e.g. `MP_OBJ_ARRAY_MVOFF(x) = 0`. All that is looking more hacky IMHO.

Reply:
I don't want to have a short macro like `memview_offset` defined in the global namespace (since it's in a header file).  So, please either put this define in objarray.c, or use an accessor macro in the header that is prefixed with `MP_`.

Reply:
> I don't want to have a short macro like memview_offset defined in the global namespace (since it's in a header file).

I see your concern now. Well, this file isn't supposed to be part of "global namespace", but rather be included by "friend modules", interested accessing internal structure of array objects. But yes, grepping thru the entire tree, it's included in a number of places, so again, I see the concern.

But that also helps to explain the concern with `MP_OBJ_ARRAY_MVOFF(x)` - it kinda introduces "semi-public API". But that's not the intention of this patch, it just tries to fix implementation issue. Leaving more general solution (which will affect "semi-public" API) for later.

So, my preference would be to move this define to objarray.c indeed.